AudioDeviceExample/AudioDevices/ExampleAudioEngineDevice.m

```diff
@@ -0,0 +1,741 @@
+                                                               UInt32 audioBufferSizeInBytes = bufferList->mBuffers[0].mDataByteSize;
+
+                                                               /*
+                                                                * Pull decoded the all remote Participant's mixed audio data from the media
+                                                                * engine into the AudioUnit's AudioBufferList.
+                                                                */
+                                                               TVIAudioDeviceReadRenderData(context->deviceContext, audioBuffer, audioBufferSizeInBytes);
+
+                                                               return bufferList;
+                                                           }];
+        if (!success) {
+            NSLog(@"Failed to set the manual rendering block");
+            return;
+        }
+
+        success = [_engine startAndReturnError:&error];
+        if (!success) {
+            NSLog(@"Failed to start AVAudioEngine, error = %@", error);
+        }
+    }
+}
+
+- (void)teardownAudioEngine {
+    [_engine stop];
+    _engine = nil;
+}
+
+- (void)playMusic {
+    if (!_engine) {
+        NSLog(@"Cannot play music. AudioEngine has not been created yet.");
+        return;
+    }
+
+    if (_player.isPlaying) {
+        [_player stop];
+    }
+
+    NSString *fileName = [NSString stringWithFormat:@"%@/%@", [[NSBundle mainBundle] bundlePath], @"mixLoop.caf"];
+    NSURL *url = [NSURL fileURLWithPath:fileName];
+    AVAudioFile *file = [[AVAudioFile alloc] initForReading:url error:nil];
+
+    _player = [[AVAudioPlayerNode alloc] init];
+    _reverb = [[AVAudioUnitReverb alloc] init];
+
+    [_reverb loadFactoryPreset:AVAudioUnitReverbPresetMediumHall];
+    _reverb.wetDryMix = 50;
+
+    [_engine attachNode:_player];
+    [_engine attachNode:_reverb];
+
+    [_engine connect:_player to:_reverb format:file.processingFormat];
+    [_engine connect:_reverb to:_engine.mainMixerNode format:file.processingFormat];
+
+    [_player scheduleFile:file atTime:nil completionHandler:^{}];
+    [_player play];
+}
+
+#pragma mark - TVIAudioDeviceRenderer
+
+- (nullable TVIAudioFormat *)renderFormat {
+    if (!_renderingFormat) {
+
+        if (!self.isCapturerInitialized) {
+            [self setupAVAudioSession];
+            _rendererInitialized = YES;
+        }
+
+        /*
+         * Assume that the AVAudioSession has already been configured and started and that the values
+         * for sampleRate and IOBufferDuration are final.
+         */
+        _renderingFormat = [[self class] activeFormat];
+    }
+
+    return _renderingFormat;
+}
+
+- (BOOL)initializeRenderer {
+    /*
+     * In this example we don't need any fixed size buffers or other pre-allocated resources. We will simply write
+     * directly to the AudioBufferList provided in the AudioUnit's rendering callback.
+     */
```

Comment:
Does this apply to the new class? It looked like you were providing a buffer.

Edit: Looks like this is still true for rendering. 👍